File: iOS11bookExamples/bk2ch08p467collectionViewLayoutFromScratchObjc/ch21p748collectionViewFlowLayout2/MyLayout.m


#import "MyLayout.h"

@implementation MyLayout {
    NSMutableArray* _atts;
    CGSize _sz;
}

-(void)prepareLayout {
    // how many items are there in total?
    int total = 0;
    NSInteger sections = [self.collectionView numberOfSections];
    for (int i = 0; i < sections; i++)
        total += [self.collectionView numberOfItemsInSection:i];
    
    // work out cell size based on bounds size
    CGSize sz = self.collectionView.bounds.size;
    CGFloat width = sz.width;
    int shortside = floor(width/50.0);
    CGFloat cellside = width/(float)shortside;
    
    // generate attributes for all cells
    int x = 0;
    int y = 0;
    NSMutableArray* atts = [NSMutableArray new];
    for (int i = 0; i < sections; i++) {
        NSInteger jj = [self.collectionView numberOfItemsInSection:i];
        for (int j = 0; j < jj; j++) {
            UICollectionViewLayoutAttributes* att =
            [UICollectionViewLayoutAttributes
             layoutAttributesForCellWithIndexPath:
             [NSIndexPath indexPathForItem:j
                                 inSection:i]];
            att.frame = CGRectMake(x*cellside,y*cellside,cellside,cellside);
            [atts addObject:att];
            x++;
            if (x >= shortside) {
                x = 0;
                y++;
            }
        }
    }
    self->_atts = atts;
    int fluff = (x == 0) ? 0 : 1;
    self->_sz = CGSizeMake(width, (y+fluff) * cellside);
}

- (CGSize)collectionViewContentSize {
    return self->_sz;
}

-(BOOL)shouldInvalidateLayoutForBoundsChange:(CGRect)newBounds {
    return newBounds.size.width != self->_sz.width;
}

- (UICollectionViewLayoutAttributes *)layoutAttributesForItemAtIndexPath:(NSIndexPath *)indexPath {
    for (UICollectionViewLayoutAttributes* att in self->_atts) {
        if ([att.indexPath isEqual:indexPath])
            return att;
    }
    return nil; // shouldn't happen
}

- (NSArray *)layoutAttributesForElementsInRect:(CGRect)rect {
    return self->_atts;
}

@end

File: bk2ch08p467collectionViewLayoutFromScratchSwift/ch21p748collectionViewFlowLayout2/MyLayout.swift

import UIKit

class MyLayout : UICollectionViewLayout {
    
    var sz = CGSize.zero
    var atts = [IndexPath:UICollectionViewLayoutAttributes]()
    
    // absolute rock-bottom layout from scratch, shows minimal responsibilities
    
    override func prepare() {
        print("prepare")
        let sections = self.collectionView!.numberOfSections()
        
        /*
        // how many items are there in total?
        let total = Array(0 ..< sections).map {
            self.collectionView!.numberOfItemsInSection($0)
            }.reduce(0, combine: +)
        */
        
        // work out cell size based on bounds size
        let sz = self.collectionView!.bounds.size
        let width = sz.width
        let shortside = floor(width/50.0)
        let cellside = width/shortside
        
        // generate attributes for all cells
        var x = 0
        var y = 0
        var atts = [UICollectionViewLayoutAttributes]()
        for i in 0 ..< sections {
            let jj = self.collectionView!.numberOfItems(inSection:i)
            for j in 0 ..< jj {
                let att = UICollectionViewLayoutAttributes(
                    forCellWith:
                    IndexPath(item:j, section:i))
                att.frame = CGRect(CGFloat(x)*cellside,CGFloat(y)*cellside,cellside,cellside)
                atts += [att]
                x += 1
                if CGFloat(x) >= shortside {
                    x = 0
                    y += 1
                }
            }
        }
        for att in atts {
            self.atts[att.indexPath] = att
        }
        let fluff = (x == 0) ? 0 : 1
        self.sz = CGSize(width, CGFloat(y+fluff) * cellside)
    }
    
    override func collectionViewContentSize() -> CGSize {
        //        println("size")
        return self.sz
    }
    
    override func shouldInvalidateLayout(forBoundsChange newBounds: CGRect) -> Bool {
        let ok = newBounds.size.width != self.sz.width
        print("should \(ok)")
        return ok
    }
    
    override func layoutAttributesForItem(at indexPath: IndexPath) -> UICollectionViewLayoutAttributes? {
        print("atts")
        return self.atts[indexPath]
    }
    
    override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
        //        println("rect")
        return Array(self.atts.values)
    }
}
